[Windows] address recommendations of VS code analysis tools for WinMain

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Code analysis issues for WinMain:
```
WinMain.cpp(40): warning C28251: Inconsistent annotation for 'WinMain': this instance has no annotations. See c:\program files (x86)\windows kits\10\include\10.0.22621.0\um\winbase.h(1037). 
WinMain.cpp(102): warning C6031: Return value ignored: 'CoInitializeEx'.
WinMain.cpp(106): warning C6031: Return value ignored: 'WSAStartup'.
WinMain.cpp(132): warning C6387: 'appRunningMutex' could be '0'.
```
The errors are raised before logging facilities are initialized. Changing start order would be a big multiplatform undertaking, so simply output to debugger if attached, and to a console/cmd.exe that started Kodi.

Adding own console for a cleaner output is a lot more work for situations that will almost never happen. For those rare cases we'll be able to ask the user to start kodi from cmd.exe and to get the exact message and error code.
There are convenient Win API methods for debug mode but they are commented out in release mode so useless here.

Return code values: I didn't find a place where Kodi defines the possible application return codes, so started at -1 and went downwards. It may conflict with other values in use.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Low priority code issues, still worth taking care of. The annotations are annoying with no impact.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Artificially raised the errors in code, under debugger and when started from cmd.exe

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
More logging for very rare cases of Kodi not getting to the point of starting the core.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[Windows] Address recommendations of VS code analysis tools for WinMain

annotations of WinMain
return code of CoInitializeEx and WSAStartup not checked

thexai commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "platform/xbmc.h"
 #include "threads/Thread.h"
 #include "utils/CharsetConverter.h" // Required to initialize converters before usage
+#include "utils/StringUtils.h"
```
I'm not sure of include `StringUtils` here: is a big code and if produces crash at this point is hard to debug as main kodi logs, mini dump, etc is not yet initialized. Maybe use simple wsprintf or system equivalent for formatting these strings.

CrystalP replied:
ok. took a while to find the correct variation of printf for safe buffer allocation but I have something.

thexai commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -84,26 +99,43 @@ INT WINAPI WinMain(HINSTANCE hInst, HINSTANCE, LPSTR commandLine, INT)
   std::string appName = CCompileInfo::GetAppName();
   HANDLE appRunningMutex = CreateMutex(nullptr, FALSE, ToW(appName + " Media Center").c_str());
-  if (GetLastError() == ERROR_ALREADY_EXISTS)
+  if (appRunningMutex == nullptr)
```
This code change is entirely wrong to check `GetLastError() == ERROR_ALREADY_EXISTS` appRunningMutex is valid HANDLE then is not possible `appRunningMutex == nullptr` and `GetLastError() == ERROR_ALREADY_EXISTS` at same time.

CreateMutex succeeds (!nullptr) and GetLastError() == ERROR_ALREADY_EXISTS when other instance of Kodi is already running.

CrystalP replied:
Good call, I switched a couple things. Also noticed that ReleaseMutex at the end returns in error, because mutex ownership is not claimed during creation and it's not waited on either. I'll remove.
What's needed instead is a CloseHandle().

thexai commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -84,26 +99,43 @@ INT WINAPI WinMain(HINSTANCE hInst, HINSTANCE, LPSTR commandLine, INT)
+      LogError(L"unable to create unique instance mutex");
+      return -1;
     }
-    ReleaseMutex(appRunningMutex);
```
When exit early (because another instance is already running) is correct ReleaseMutex

thexai commented:
As a general comment: what IntelliSense suggests will not always make sense and when it does it is usually unimportant things. For example, it is not necessary to always use the return values of all functions in all calls, just because the function returns a value.

CrystalP committed:
[Windows] Close the application mutex when exiting

ReleaseMutex was incorrect because the mutex ownership is not requested and it's not waited on.
Instead the handle should be closed.

CrystalP commented:
Technically could go further to ensure that WSACleanup / CoUninitialize / CloseHandle(appRunningMutex) / delete[] argv[i] get called for every "return" path but since the app is closing, all resources will be released to the OS anyway and it's probably not worth the effort.

CrystalP committed:
[Windows] Uninitialize properly for all exit paths

CrystalP committed:
[Windows] Extract command line parsing into its own function

thexai commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -33,16 +33,38 @@ LONG WINAPI CreateMiniDump(EXCEPTION_POINTERS* pEp)
+template<typename... Args>
+static void LogError(const wchar_t* format, Args&&... args)
+{
+  int count = _snwprintf(nullptr, 0, format, args...);
```
```suggestion
  const int count = _snwprintf(nullptr, 0, format, args...);
```

thexai commented on xbmc/platform/win32/WinMain.cpp:
```diff
@@ -116,20 +166,21 @@ INT WINAPI WinMain(HINSTANCE hInst, HINSTANCE, LPSTR commandLine, INT)
-  ReleaseMutex(appRunningMutex);
+cleanup:
+
+  if (!rcWinsock)
```
```suggestion
  if (rcWinsock == 0)
```

CrystalP commented:
Applied both suggestions

CrystalP commented:
I doubt android and webos build failures are related to WinMain changes, pushing the button :-)
